### DISCUSSION TITLE
Inject settings thru environment variables

### DISCUSSION THREAD

**Discussion by sorbonad** [dotnet/yarp#2169] [Q&A] [answered]
Hi! Should be possible to load the reverse proxy configuration from a environment variable instead of relying on the appsettings.json?

I understand that I can override some parts by using config layering but I am trying to keep the whole config structure out in a env var.

Thanks!

**Comment by Tratcher** [comment]
Yes, any config value can be represented as an environment variable, but you would need a lot of them.
https://learn.microsoft.com/aspnet/core/fundamentals/configuration/#naming-of-environment-variables

**Reply by sorbonad** [reply]
Thanks for the reply.  Right, I was thinking of saving the full configuration as json in one environment variable and later loading it on startup. Would that be possible without creating a custom config provider?

**Reply by Tratcher** [reply]
Hmm, I don't know how well that would work, I'd expect to hit syntax conflicts.

Can you share why you're trying to do this exclusively from Env Vars? There may be other options.

**Reply by sorbonad** [reply]
Sure, in my case I need to deploy instances of YARP which will run in isolated from each other in different clusters. The settings of those instances will be different and can independently vary as time goes by. The current deployment strategy assumes that you can set settings thru environment variables. I need to have only one central 'vanilla' image. Currently the deployment pipeline injects the proper configuration by using environment variables.

**Reply by Tratcher** [reply]
Env vars can't be changed without restarting the app. What about an env var that pointed to a scenario specific config file? Or having the deployment pipeline inject a specific config file? It already seems to know the entire config.

**Reply by sorbonad** [reply]
Those were some options I was about to try, I will do some test and I will let you know. Thanks again!

**Reply by sorbonad** [reply, marked as answer]
Ater browsing the code a little bit, I believe this would work:

```c#
var builder = WebApplication.CreateBuilder(args);
builder.Configuration.AddJsonStream(new MemoryStream(Encoding.UTF8.GetBytes(GetJsonConfigFromEnvironmentVariables())));
...
```  

GetJsonConfigFromEnvironmentVariables is just reading the env variable which contains a valid json config payload.

I believe this won't play nicely with the auto config hot reloads, but in my case that scenario is not going to happen since every configuration/code change requires a deployment. 

What do you think?

**Reply by Tratcher** [reply]
Seems workable. It's a shame not to support hot reloads though.

**Reply by sorbonad** [reply]
I agree, I will revisit that approach when possible.
